src/python/pants/option/option_value_container.py

```diff
@@ -108,25 +108,25 @@ def is_default(self, key: Key) -> bool:
         """
         return self.get_rank(key) in (Rank.NONE, Rank.HARDCODED)
 
-    def get(self, key: Key, default: Value | None = None):
+    def get(self, key: Key, default: Value | None = None, throw_errors: bool = False):
         # Support dict-like dynamic access.  See also __getitem__ below.
         if key not in self._value_map:
             return default
-        return self._get_underlying_value(key)
+        return self._get_underlying_value(key, throw_errors)
 
     def as_dict(self) -> dict[Key, Value]:
         return {key: self.get(key) for key in self._value_map}
 
     def to_builder(self) -> OptionValueContainerBuilder:
         return OptionValueContainerBuilder(copy.copy(self._value_map))
 
-    def _get_underlying_value(self, key: Key):
+    def _get_underlying_value(self, key: Key, throw_errors: bool):
         # Note that the key may exist with a value of None, so we can't just
         # test self._value_map.get() for None.
         if key not in self._value_map:
             raise AttributeError(key)
         ranked_val = self._value_map[key]
-        if ranked_val.error is not None:
+        if throw_errors and ranked_val.error is not None:
```

Comment:
I'm not crazy about this - it is very hard to reason about which errors are possible here, and what the consequence is of ignoring each one of them - what the value might be, and what happens downstream, etc.

How would you feel about reverting the required option functionality? As far as I can tell we have very few required options, so it's not clear that the feature is worth this added complication.

I see required options as just one special case of option validation - and in our case not even a common one. Other kinds of option validation include "if this option is specified then this other one must not be"  or "if this option has this value then this other option must be present", and so on.   And since all those must be enforced manually, I'm not sure it's worth carving out a special affordance for that one special case?

Reply:
If we can address the concerns, that would be preferable imo. Failing that I'm open to revert the change. The motivation was to get rid boiler platey code -- but I see an alternative approach to solve that where we may change how we look up the value rather than enforce the general path. 

So yea, I'd be OK with reimplementing how the required option is done.

--
This was spurred from me looking at adding yet another goal that would have "required" options (as arguments are for specs/goals only) and wasn't too fond of the explicit checks for each option. Lots of duplication and the error message will be hard to keep consistent across goals over time.

Reply:
The options code is the oldest remaining part of the codebase (and basically the only holdover from v1). It's already very complex, so I am leery of changes like this or like the part of #18780 that adds the `error` field in ranked_value.py. I fear that they make an already complicated system even harder to reason about.

How about if we revert the changes to ranked_value.py in #18780, and the `Missing` stuff in option_types.py, but keep the new `required=` keyword, and validate it when we parse, rather than lazily when we read. 

That has the added advantage that we can tell you about all the missing required options in one go, instead of one at a time.

Reply:
I guess there's two "classes" of required options here. Ones that make sense to enforce always, that would work well with what you describe, while other that may only be required as inputs to a certain goal which doesn't make sense to always enforce but only during that goals execution (or really, whenever those options are accessed).